gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/ApiServiceImpl.java

```diff
@@ -3092,4 +3093,55 @@ private void triggerNotification(String apiId, ApiHook hook, ApiEntity apiEntity
             notifierService.trigger(hook, apiId, new NotificationParamsBuilder().api(apiEntity).user(userService.findById(userId)).build());
         }
     }
+
+    @Override
+    public Map<String, Long> countPublishedByUserGroupedByCategories(String userId) {
+        ApiCriteria criteria = new ApiCriteria.Builder().visibility(PUBLIC).lifecycleStates(List.of(ApiLifecycleState.PUBLISHED)).build();
+        ApiFieldInclusionFilter filter = ApiFieldInclusionFilter.builder().includeCategories().build();
+
+        Set<Api> apis = apiRepository.search(criteria, filter);
+
+        final String[] userApiIds = membershipService
+            .getMembershipsByMemberAndReference(MembershipMemberType.USER, userId, MembershipReferenceType.API)
+            .stream()
+            .map(MembershipEntity::getReferenceId)
+            .toArray(String[]::new);
+
+        if (userApiIds.length > 0) {
+            criteria = new ApiCriteria.Builder().ids(userApiIds).build();
+            Set<Api> userApis = apiRepository.search(criteria, filter);
+            apis.addAll(userApis);
+        }
+
+        final String[] groupIds = membershipService
+            .getMembershipsByMemberAndReference(MembershipMemberType.USER, userId, MembershipReferenceType.GROUP)
+            .stream()
+            .filter(
+                m -> {
+                    final RoleEntity roleInGroup = roleService.findById(m.getRoleId());
+                    return m.getRoleId() != null && roleInGroup.getScope().equals(RoleScope.API);
+                }
+            )
+            .map(MembershipEntity::getReferenceId)
+            .toArray(String[]::new);
+
+        if (groupIds.length > 0 && groupIds[0] != null) {
+            criteria = new ApiCriteria.Builder().groups(groupIds).build();
+            Set<Api> groupApis = apiRepository.search(criteria, filter);
+            apis.addAll(groupApis);
+        }
+
+        Set<String> categoryIds = apis
+            .stream()
+            .map(Api::getCategories)
+            .filter(Objects::nonNull)
+            .flatMap(Set::stream)
+            .collect(Collectors.toSet());
```

Comment:
Suggestion : could we use a toMap collector here, to improve readability ?

Reply:
👍